### ISSUE TITLE
[stable/jenkins] Backup issue with serviceaccount and IAM roles

### ISSUE THREAD

**Issue by KrzesloSzatan** [helm/charts#22670]
Im deploying jenkins with terraform on AWS using jenkins-values.yaml and while i am enabling the backup 

```
backup:
  enabled: true
  componentName: "backup"
  schedule: "38 13 * * *"
  annotations: {}
  image:
    repository: "maorfr/kube-tasks"
    tag: "0.2.0"
  extraArgs: []
  existingSecret: {}
  env:
  - name: "AWS_REGION"
    value: "eu-central-1"
  resources:
    requests:
      memory: 1Gi
      cpu: 1
    limits:
      memory: 1Gi
      cpu: 1
  destination: "s3://xxx"
```
the template creates a serviceaccount jenkins-backup which then is used by the cronjob in created pods to do the backup as intended. I would like to attach a IAM role to it using the annotation as in AWS [doc](https://docs.amazonaws.cn/en_us/eks/latest/userguide/create-service-account-iam-policy-and-role.html) BUT, the problem is that i actually cannot add any annotation to this account using the template (i can do it only by using the kubectl annotate sa.... ) because there is no possible option to annotate this account in the yaml file. This annotations should then be converted "automagically" via some webhook on AWS (didnt get into details at that to cyte how this works) to proper envs for the pod

AWS_ROLE_ARN
arn:aws:iam::11111111111:role/xxxxxxxx
AWS_WEB_IDENTITY_TOKEN_FILE
/var/run/secrets/eks.amazonaws.com/serviceaccount/token

 I tried to set the annotations to the backup part like that 

```
backup:
  enabled: true
  componentName: "backup"
  schedule: "38 13 * * *"
  annotations: 
    eks.amazonaws.com/role-arn: arn:aws:iam::11111111111:role/xxxxxxxx
  image:
    repository: "maorfr/kube-tasks"
    tag: "0.2.0"
```
BUT as i have checked in the dashboard pod gets only the part "eks.amazonaws.com/role-arn:" and the rest is ommited but while i click on the edit button in the dashboard in the pod section it shows the annotation propelry but while there is a number for the tennant ID it marks the annotation with ' ' so it looks like that 

eks.amazonaws.com/role-arn: 'arn:aws:iam::11111111111:role/xxxxxxxx'

while when i change it to whatever but without the tennant id 1111111111111 it doesnt set ' ' and i can see the whole annotation in the dashboard (but not the proper one so the envs are missing). 

It works properly when i add this annotation to the jenkins-backup serviceaccount manualy using kubectl. Then it attaches proper envs to the pod BUT there is still an issue with access rights

```
WebIdentityErr: unable to read file at /var/run/secrets/eks.amazonaws.com/serviceaccount/token
caused by: open /var/run/secrets/eks.amazonaws.com/serviceaccount/token: permission denied
```

This can be fixed by editing pod and adding to the pod manually via kubectl. 

```
securityContext:
  fsGroup: 65534
```

To sum it up:
I cant add annotation to serviceaccount "jenkins-backup" created by the default backup template from yaml that should pass on the vars to the pod created by cronjob for the iam role to be used to send the backup to the bucket and i have to set fsGroup: 65534 manualy on the pod for the serviceaccount to be able to read the token. 

If I set the annotation to the jenkins-backup serviceaccount and add the fsGroup: 65534 to the pod's securitycontext manualy via kubect the backup works fine. 

Can you please point me if there is something im doing wrong or it shouldnt work like that or what is wrong ?
 
**Which chart**:
stable/jenkins

Is there anything id should mention more ? Im still learning so sry if i dint mention something crucial.

**Comment by torstenwalter** [comment]
@KrzesloSzatan Looks like it wasn't used used like this before. Could you create a PR which adds the functionality you are looking for?

Annotation for the service account could be added in https://github.com/helm/charts/blob/master/stable/jenkins/templates/jenkins-backup-rbac.yaml

Necessary values should be placed below `.Values.backup`

**Comment by njtman** [comment]
FWIW, I ran into similar issues when trying to implement the built in backup mechanism.

Aside from the `role-arn` annotation not being supported by the service account template in the Jenkins helm chart, the `aws go sdk` version included inside the `nuvo/kube-tasks` image is outdated and does not support [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html).  I suspect it only worked for you because you have a node role attached which provides access to the S3 bucket.

See [this page](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html) for minimum supported AWS SDK versions. [Here](https://github.com/nuvo/kube-tasks/blob/master/Gopkg.lock#L63) is the version of the `aws go sdk` that `nuvo/kube-tasks` is using.

**Comment by KrzesloSzatan** [comment]
Yes, ive got a role with proper access to s3 attached but afaik this is how it should be ? Anyway, ive edited the template to add the annotations to jenkins-backup serviceaccount and securitycontext and i have tested it with runAsUser: 1000 and fsGroup: 1000 and everything works. So i will try to create a PR tommorow probably when im 100% sure it works as intended.

**Comment by KrzesloSzatan** [comment]
> 
> 
> FWIW, I ran into similar issues when trying to implement the built in backup mechanism.
> 
> Aside from the `role-arn` annotation not being supported by the service account template in the Jenkins helm chart, the `aws go sdk` version included inside the `nuvo/kube-tasks` image is outdated and does not support [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html). I suspect it only worked for you because you have a node role attached which provides access to the S3 bucket, or the bucket permissions aren't correctly configured.
> 
> See [this page](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html) for minimum supported AWS SDK versions. [Here](https://github.com/nuvo/kube-tasks/blob/master/Gopkg.lock#L63) is the version of the `aws go sdk` that `nuvo/kube-tasks` is using.

**Comment by njtman** [comment]
> > FWIW, I ran into similar issues when trying to implement the built in backup mechanism.
> > Aside from the `role-arn` annotation not being supported by the service account template in the Jenkins helm chart, the `aws go sdk` version included inside the `nuvo/kube-tasks` image is outdated and does not support [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html). I suspect it only worked for you because you have a node role attached which provides access to the S3 bucket, or the bucket permissions aren't correctly configured.
> > See [this page](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html) for minimum supported AWS SDK versions. [Here](https://github.com/nuvo/kube-tasks/blob/master/Gopkg.lock#L63) is the version of the `aws go sdk` that `nuvo/kube-tasks` is using.
> 
> Yes, ive got a role with proper access to s3 attached but afaik this is how it should be ? Anyway, ive edited the template to add the annotations to jenkins-backup serviceaccount and securitycontext and i have tested it with runAsUser: 1000 and fsGroup: 1000 and everything works. So i will try to create a PR tommorow probably when im 100% sure it works as intended.

No, this is not how it should be.  The purpose of IRSA is to not have to use node roles.  Assigning AWS resource access to an individual pod vs a cluster node is more secure.  Only one pod will have access to said AWS resource instead of all pods running on the cluster node.  [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) is the natively supported EKS mechanism that enables this functionality.  Previously you had to use something like kube2iam to get this working.

**Comment by KrzesloSzatan** [comment]
> 
> 
> > > FWIW, I ran into similar issues when trying to implement the built in backup mechanism.
> > > Aside from the `role-arn` annotation not being supported by the service account template in the Jenkins helm chart, the `aws go sdk` version included inside the `nuvo/kube-tasks` image is outdated and does not support [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html). I suspect it only worked for you because you have a node role attached which provides access to the S3 bucket, or the bucket permissions aren't correctly configured.
> > > See [this page](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html) for minimum supported AWS SDK versions. [Here](https://github.com/nuvo/kube-tasks/blob/master/Gopkg.lock#L63) is the version of the `aws go sdk` that `nuvo/kube-tasks` is using.
> > 
> > 
> > Yes, ive got a role with proper access to s3 attached but afaik this is how it should be ? Anyway, ive edited the template to add the annotations to jenkins-backup serviceaccount and securitycontext and i have tested it with runAsUser: 1000 and fsGroup: 1000 and everything works. So i will try to create a PR tommorow probably when im 100% sure it works as intended.
> 
> No, this is not how it should be. The purpose of IRSA is to not have to use node roles. Assigning AWS resource access to an individual pod vs a cluster node is more secure. Only one pod will have access to said AWS resource instead of all pods running on the cluster node. [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) is the natively supported EKS mechanism that enables this functionality. Previously you had to use something like kube2iam to get this working.
> 
> If you are okay from a security perspective using node roles, nothing else is required. You wouldn't need to associate an rbac service account with a policy.

I must say im a little bit confused, i am creating a separate IAM role for the service account, attaching  policy with proper access rights to s3 to that role, then with service account annotation it creates a pod with proper envs to get a token that will enable this pod to connect to s3. That is how i understand this - doesnt have to be right ;) But to be honest i have no idea what im doing wrong from your perspective ?

**Comment by KrzesloSzatan** [comment]
> the aws go sdk version included inside the nuvo/kube-tasks image is outdated and does not support IRSA. 

This might be a clue for me :)

**Comment by njtman** [comment]
> > > > FWIW, I ran into similar issues when trying to implement the built in backup mechanism.
> > > > Aside from the `role-arn` annotation not being supported by the service account template in the Jenkins helm chart, the `aws go sdk` version included inside the `nuvo/kube-tasks` image is outdated and does not support [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html). I suspect it only worked for you because you have a node role attached which provides access to the S3 bucket, or the bucket permissions aren't correctly configured.
> > > > See [this page](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html) for minimum supported AWS SDK versions. [Here](https://github.com/nuvo/kube-tasks/blob/master/Gopkg.lock#L63) is the version of the `aws go sdk` that `nuvo/kube-tasks` is using.
> > > 
> > > 
> > > Yes, ive got a role with proper access to s3 attached but afaik this is how it should be ? Anyway, ive edited the template to add the annotations to jenkins-backup serviceaccount and securitycontext and i have tested it with runAsUser: 1000 and fsGroup: 1000 and everything works. So i will try to create a PR tommorow probably when im 100% sure it works as intended.
> > 
> > 
> > No, this is not how it should be. The purpose of IRSA is to not have to use node roles. Assigning AWS resource access to an individual pod vs a cluster node is more secure. Only one pod will have access to said AWS resource instead of all pods running on the cluster node. [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) is the natively supported EKS mechanism that enables this functionality. Previously you had to use something like kube2iam to get this working.
> > If you are okay from a security perspective using node roles, nothing else is required. You wouldn't need to associate an rbac service account with a policy.
> 
> I must say im a little bit confused, i am creating a separate IAM role for the service account, attaching policy with proper access rights to s3 to that role, then with service account annotation it creates a pod with proper envs to get a token that will enable this pod to connect to s3. That is how i understand this - doesnt have to be right ;) But to be honest i have no idea what im doing wrong from your perspective ?

Based on what you just said, yes, your understanding is correct.  When you said "Yes, ive got a role with proper access to s3 attached", I thought you meant you were using a node role in addition to IRSA.  If you are only using IRSA, access to the S3 bucket by the backup container should not work because the outdated `aws go sdk` version doesn't support IRSA.

I tested this myself about a month or two ago and ran into this very issue.  If the backup container can in fact successfully access the S3 bucket after configuring IRSA, I suspect something else is at play here (a node role or lack of restriction on the s3 bucket).

**Comment by KrzesloSzatan** [comment]
> 
> 
> > > > > FWIW, I ran into similar issues when trying to implement the built in backup mechanism.
> > > > > Aside from the `role-arn` annotation not being supported by the service account template in the Jenkins helm chart, the `aws go sdk` version included inside the `nuvo/kube-tasks` image is outdated and does not support [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html). I suspect it only worked for you because you have a node role attached which provides access to the S3 bucket, or the bucket permissions aren't correctly configured.
> > > > > See [this page](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html) for minimum supported AWS SDK versions. [Here](https://github.com/nuvo/kube-tasks/blob/master/Gopkg.lock#L63) is the version of the `aws go sdk` that `nuvo/kube-tasks` is using.
> > > > 
> > > > 
> > > > Yes, ive got a role with proper access to s3 attached but afaik this is how it should be ? Anyway, ive edited the template to add the annotations to jenkins-backup serviceaccount and securitycontext and i have tested it with runAsUser: 1000 and fsGroup: 1000 and everything works. So i will try to create a PR tommorow probably when im 100% sure it works as intended.
> > > 
> > > 
> > > No, this is not how it should be. The purpose of IRSA is to not have to use node roles. Assigning AWS resource access to an individual pod vs a cluster node is more secure. Only one pod will have access to said AWS resource instead of all pods running on the cluster node. [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) is the natively supported EKS mechanism that enables this functionality. Previously you had to use something like kube2iam to get this working.
> > > If you are okay from a security perspective using node roles, nothing else is required. You wouldn't need to associate an rbac service account with a policy.
> > 
> > 
> > I must say im a little bit confused, i am creating a separate IAM role for the service account, attaching policy with proper access rights to s3 to that role, then with service account annotation it creates a pod with proper envs to get a token that will enable this pod to connect to s3. That is how i understand this - doesnt have to be right ;) But to be honest i have no idea what im doing wrong from your perspective ?
> 
> Based on what you just said, yes, your understanding is correct. When you said "Yes, ive got a role with proper access to s3 attached", I thought you meant you were also using a node role in addition to IRSA. If you are only using IRSA, access to the S3 bucket by the backup container should not work because the outdated `aws go sdk` version doesn't support IRSA.
> 
> I tested this myself about a month or two ago and ran into this very issue. If the backup container can in fact successfully access the S3 bucket after configuring IRSA, I suspect something else is at play here (a node role or lack of restriction on the s3 bucket).

I will check the bucket restrictions tomorrow (its midnight here ;) ) cuz ive got a feeling that there might be a leak, like you said...

**Comment by cmcga1125** [comment]
I know this ticket was started regarding AWS eks, but having the ability to add an annotation to the service account would also fix an issue I'm having with GCP GKE [Workload Identity.](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity)

is there any movement on this? Could it be as easy as mapping in the annotations from the pod? or would it be a better idea to make a specified key/value?

**Comment by KrzesloSzatan** [comment]
> 
> 
> I know this ticket was started regarding AWS eks, but having the ability to add an annotation to the service account would also fix an issue I'm having with GCP GKE [Workload Identity.](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity)
> 
> is there any movement on this? Could it be as easy as mapping in the annotations from the pod? or would it be a better idea to make a specified key/value?

I've posted a PR for that, but dunno when or if it will be merged. Just for now ive made a local chart for myself and use it as i changed it for my needs. I think you can do the same for the time beeing. You can take whatever you want from my fork https://github.com/KrzesloSzatan/charts/tree/master/stable/jenkins#backup

**Comment by torstenwalter** [comment]
@KrzesloSzatan can you send a link to that PR? I'll have a look.

**Comment by cmcga1125** [comment]
currently, I'm using our CICD tool to just annotate the SA after deployment. Which is fine, but can send a PR if @KrzesloSzatan doesn't

**Comment by KrzesloSzatan** [comment]
@torstenwalter Sry, i was sure i send the link already. https://github.com/helm/charts/pull/22730 you have already commented on that, but im trying to do as instructed (im not that fluent in git unfortunately) and have done the --signoff few times already but nothing seem to change
